build/source/engine/allocspace.f90

```diff
@@ -210,33 +210,49 @@ subroutine allocGlobal(metaStruct,dataStruct,err,message)
  end do
 
  ! * allocate local data structures where there is a spatial dimension
- do iGRU=1,nGRU
+ gruLoop: do iGRU=1,nGRU
+
   ! initialize the spatial flag
-  spatial=.true.
-  ! get the number of snow and soil layers
-  associate(&
-  nHRU  => gru_struc(iGRU)%hruCount,         & ! number of HRUs
-  nSnow => gru_struc(iGRU)%hruInfo(:)%nSnow, & ! number of snow layers for each HRU
-  nSoil => gru_struc(iGRU)%hruInfo(:)%nSoil  ) ! number of soil layers for each HRU
-  ! allocate space
+  spatial=.false.
+
+  ! loop through HRUs
+  hruLoop: do iHRU=1,gru_struc(iGRU)%hruCount
+
+   ! get the number of snow and soil layers
+   associate(&
+   nSnow => gru_struc(iGRU)%hruInfo(iHRU)%nSnow, & ! number of snow layers for each HRU
+   nSoil => gru_struc(iGRU)%hruInfo(iHRU)%nSoil  ) ! number of soil layers for each HRU
+
+   ! allocate space for structures WITH an HRU dimension
+   select type(dataStruct)
+    type is (gru_hru_int);       call allocLocal(metaStruct,dataStruct%gru(iGRU)%hru(iHRU),nSnow,nSoil,err,cmessage); spatial=.true.
+    type is (gru_hru_intVec);    call allocLocal(metaStruct,dataStruct%gru(iGRU)%hru(iHRU),nSnow,nSoil,err,cmessage); spatial=.true.
+    type is (gru_hru_double);    call allocLocal(metaStruct,dataStruct%gru(iGRU)%hru(iHRU),nSnow,nSoil,err,cmessage); spatial=.true.
+    type is (gru_hru_doubleVec); call allocLocal(metaStruct,dataStruct%gru(iGRU)%hru(iHRU),nSnow,nSoil,err,cmessage); spatial=.true.
```

Comment:
I am a beginner in the use of unlimited polymorphic variables -- I understand that we do need this `select type` construct: Specifically, since `dataStruct` is polymorphic, we need case-appropriate statements. True, the calls to `allocLocal` are identical, though all are needed in this context. Others can test alternative constructs, though I'll stick with this one for now.